Make comments panel collapsible #4794

tarun14110 opened the pull request:
Fixes #4794 

![screenshot from 2016-02-19 22 29 04](https://cloud.githubusercontent.com/assets/12219700/13182652/c6c57bfe-d758-11e5-94f0-4b4d16a09636.png)

thyageshm commented:
@tarun14110 feel free to remind @JosephineKwa to review this pr if some time passes by

tarun14110 commented:
@thyageshm sure

tarun14110 commented:
@JosephineKwa  any improvement I need to do ?

JosephineKwa commented on src/main/webapp/js/instructor.js:
```diff
@@ -332,4 +332,14 @@ function updateHoverShowPictureEvents(actualLink, resolvedLink) {
 
+function toggleCommentsForStudents(sender){
+    var panelBody = $(sender).parent().find('div[class^="panel-body"]');
+    if(panelBody.hasClass('hidden')){
```
Rather than having this conditional, perhaps you could use the jquery method toggleClass() or even simply just use toggle() to show/hide the panel body

JosephineKwa commented:
It would also be good to add a test for this when you're done :)

JosephineKwa commented on src/main/webapp/js/feedbackResponseComments.js:
```diff
@@ -372,20 +372,27 @@ function loadFeedbackResponseComments(user, courseId, fsName, fsIndx, sender) {
-      $(sender).find('div[class^="placeholder-img-loading"]').html("");
-    });
+   
+    if(panelBody.hasClass('hidden')){
```
With regards to enabling collapse for feedback response comments, it is outside the scope of issue #4794. I believe issue #3276 should have addressed this.

tarun14110 replied:
@JosephineKwa  I looked over above issue. It seems that the collapsing of the lowers tabs were addressed and it is working finea it has been merged. Do I need to only take care of  "comments for students" tab ?

JosephineKwa replied:
Yup, the collapsing of those panels have been addressed. You only need to take care of the "comments for students" panel.

JosephineKwa replied:
Do also remember to address the comments for students panels in the other pages mentioned in the issue, namely `instructorSearch.jsp`, `instructorStudentRecords.jsp` and `studentComments.jsp`.

tarun14110 replied:
sure

JosephineKwa commented:
Apologies for the delay!

wkurniawan07 commented:
@tarun14110 if this requires a change to HTML test files (I have a high confidence it will), do sync with the master first before using God mode.

tarun14110 commented:
@JosephineKwa  sorry for delay, actually my mid-sems were going on. I had implemented the collapse/expand as you suggested. I used the same function used in instructorFeedbackResult page. I moved that function to intructor.js , to use that functon in multiple instructor pages. And for students comment page, I used a separate copy of the function. Should I move collapse/expand function to more general js file (common.js) or is it fine ?

tarun14110 commented:
@wkurniawan07 yeah, this require to change HTML test files. I will keep in mind to sync before using God mode.

thyageshm commented:
@tarun14110 please complete this soon. inform @JosephineKwa if the PR is ready for review

tarun14110 commented:
@JosephineKwa please review

JosephineKwa commented on src/main/webapp/WEB-INF/tags/instructor/search/commentsForResponsesSearchResults.tag:
```diff
@@ -6,13 +6,22 @@
 <div class="panel panel-primary">
     <div class="panel-heading">
-        <strong><jsp:doBody/></strong>
+        <strong><jsp:doBody /></strong>
```
There shouldn't be a need to change this tag, this tag is for response comments, not comments for students

JosephineKwa commented on src/main/webapp/WEB-INF/tags/instructor/search/studentsSearchResults.tag:
```diff
@@ -7,9 +7,12 @@
 <div class="panel panel-primary">
     <div class="panel-heading">
         <strong><jsp:doBody/></strong>
+        <div class='display-icon pull-right'>
```
Similarly, there shouldn't be a need to change this tag as this tag is for search results of students, rather than comments for students.

tarun14110 replied:
Sorry for these changes. Actually I was doing collapsing functionality for this before. After that realized that this was already corrected in other issue. And forgot to revert back these changes.

JosephineKwa commented on src/main/webapp/WEB-INF/tags/shared/commentsPanel.tag:
```diff
@@ -21,65 +21,89 @@
 <div class="panel panel-primary">
     <div class="panel-heading">
+        <div class='display-icon pull-right'>
+            <span class="glyphicon ${'glyphicon-chevron-up'} pull-right"></span>
```
Remove the "`${}`" surrounding the `glyphicon-chevron-up`

JosephineKwa commented on src/main/webapp/WEB-INF/tags/student/comments/feedbackSessionPanel.tag:
```diff
@@ -6,8 +6,11 @@
 <div class="panel panel-primary">
     <div class="panel-heading">
         <strong>Comments in session: ${feedbackSessionRow.feedbackSessionName}</strong>
+        <div class='display-icon pull-right'>
```
Again, there shouldn't be a need to change this tag, this tag is for feedback session panels, not comments for students panels.

tarun14110 replied:
I do added collapse/ expand functionality to comments Page of Student . Am I supposed to do this only Instructor Search Page for this issue ?

JosephineKwa replied:
On the Student comments page, there are 2 types of comments panels, the general student comments panel and the response comments panel. This tag corresponds to the response comments panel, which isn't the panel that we want to collapse (in this issue at least). Changing the commentsPanel tag is sufficient as it is the common tag that is used by the 4 pages to show the student comments panel.

JosephineKwa commented on src/main/webapp/WEB-INF/tags/shared/commentsPanel.tag:
```diff
@@ -21,65 +21,89 @@
         <strong>${bodyTitle}</strong>
     </div>
-    <div class="panel-body">
+    <div class="panel-collapse collapse in">
```
This div should contain a `panel-body` class

tarun14110 replied:
@JosephineKwa I edited collapse function for this, to be able to work on both .panel-body and .panel-collapse .  But .panel-body is not working as smooth as .panel-collapse. Should I go for .panel-body ?

tarun14110 replied:
@JosephineKwa  I will do this as soon as you will confirm.

JosephineKwa replied:
I see, perhaps you could try putting the <div class="panel-body"> element within the `<div class="panel-collapse collapse in">` element.

JosephineKwa commented on src/main/webapp/WEB-INF/tags/shared/commentsPanel.tag:
```diff
@@ -21,65 +21,89 @@
-        <c:forEach items="${commentsForStudentsTables}" var="commentsForStudentsTable">
-            <div class="panel panel-info student-record-comments${commentsForStudentsTable.extraClass}"
-                 <c:if test="${empty commentsForStudentsTable.rows && not forRecordsPage}">style="display: none;"</c:if>>
+        <c:set var="commentIndex" value="${0}" />
```
The rest of the changes below in this file all seem to be formatting changes. do revert them. If you find that other parts of the code requires reformatting or refactoring, you could create an issue for it and work on it in a separate PR.

thyageshm commented:
@tarun14110, gentle reminder to complete this by this release

JosephineKwa commented on src/main/webapp/WEB-INF/tags/shared/commentsPanel.tag:
```diff
@@ -21,65 +21,89 @@
 </c:choose>
 <div class="panel panel-primary">
     <div class="panel-heading">
```
With regards to the collapsing of the panel, perhaps using the `data-toggle="collapse"` attribute on the `panel-heading` would be sufficient. You could move the `toggleChevron()` (as well as `setChevronToUp()` and `setChevronToDown()`) method from feedbackResponseComments.js to common.js, and use it to perform the toggling of the chevron.

JosephineKwa replied:
With regards to the collapsing of the panel, perhaps using the data-toggle="collapse" attribute on the panel-heading would be sufficient, instead of using bindCollapseEvents(). You could move the toggleChevron() (as well as setChevronToUp() and setChevronToDown()) method from feedbackResponseComments.js to common.js, and use it to perform the toggling of the chevron.

tarun14110 commented:
@JosephineKwa I have updated above mentioned suggestions. please have a look.

JosephineKwa commented:
Changes look good :+1: . Do also update the test templates :)

tarun14110 commented:
@JosephineKwa have a look please.

JosephineKwa commented:
@tarun14110 , I'm getting test failures on line 156 of `InstructorCommentsPageUiTest`

tarun14110 commented:
@JosephineKwa  I am getting this error. Could you help me with this ?

```
java.lang.AssertionError: 
	at org.testng.AssertJUnit.fail(AssertJUnit.java:59)
	at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:24)
	at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:33)
	at teammates.test.cases.ui.browsertests.InstructorCommentsPageUiTest.testPanelsCollapseExpand(InstructorCommentsPageUiTest.java:156)
	at teammates.test.cases.ui.browsertests.InstructorCommentsPageUiTest.allTests(InstructorCommentsPageUiTest.java:34)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:334)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:329)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:291)
	at org.testng.SuiteRunner.run(SuiteRunner.java:240)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.testng.remote.RemoteTestNG.run(RemoteTestNG.java:111)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:204)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:175)

```

JosephineKwa commented on src/main/webapp/WEB-INF/tags/shared/commentsPanel.tag:
```diff
@@ -19,10 +19,15 @@
-<div class="panel panel-primary">
-    <div class="panel-heading">
+<div class="panel panel-primary" data-toggle="collapse" data-target="#panelBodyCollapse">
+    <div class="panel-heading cursor-pointer"
```
@tarun14110 my apologies, I had overlooked this, the data-toggle and data-target attributes should be in the `panel-heading` element rather than the `.panel` element.

JosephineKwa replied:
Having the `data-toggle` and `data-target` attributes in the `.panel` element causes the panel to collapse even when clicking on the panel body. This likely caused some of the test failures that I encountered. After making the changes, the templates will have to be updated again.

tarun14110 commented:
@JosephineKwa  Is it fine ?
But I am still getting below error in InstructorCommentPageUiTest
```
java.lang.AssertionError: 
	at org.testng.AssertJUnit.fail(AssertJUnit.java:59)
	at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:24)
	at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:33)
	at teammates.test.cases.ui.browsertests.InstructorCommentsPageUiTest.testPanelsCollapseExpand(InstructorCommentsPageUiTest.java:156)
	at teammates.test.cases.ui.browsertests.InstructorCommentsPageUiTest.allTests(InstructorCommentsPageUiTest.java:34)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

```

JosephineKwa commented:
line 156 of `InstructorCommentsPageUiTest` checks if all the `.panel-collapse` elements are hidden when `InstructorCommentsPage` is loaded. Originally only the comments in session panels had the `.panel-collapse` class, but after the changes that you made, the student comments panel also has the `.panel-collapse class`. Hence when the check is made, it sees that the student comments panel is visible and thus fails. You will need to modify the `testPanelsCollapseExpand` method to take into account the changes that you made. This is likely not the only method failing due to the changes in this PR, thus you will need to check for test failures in other tests and update them as necessary in order to get dev green.

Devmassnk commented:
@JosephineKwa  I changed the InstructorCommentPageUiTests. 
Inside testPanelsCollapseExpand() , I first clicked the 'comments for students' panel to collapse that. And all others were as they were. 
But I am randomly getting failed test. I think issue is with waitForPanelsToCollapse(); because on introducing some delay before `assertTrue(commentsPage.areCommentsHidden());` makes test pass consistently. Could you help me with this ?

`java.lang.AssertionError: 
	at org.testng.AssertJUnit.fail(AssertJUnit.java:59)
	at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:24)
	at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:33)
	at teammates.test.cases.ui.browsertests.InstructorCommentsPageUiTest.testPanelsCollapseExpand(InstructorCommentsPageUiTest.java:167)
	at teammates.test.cases.ui.browsertests.InstructorCommentsPageUiTest.allTests(InstructorCommentsPageUiTest.java:34)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:334)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:329)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:291)
	at org.testng.SuiteRunner.run(SuiteRunner.java:240)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.testng.remote.RemoteTestNG.run(RemoteTestNG.java:111)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:204)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:175)
`

tarun14110 commented:
@JosephineKwa I changed the InstructorCommentPageUiTests. 
Inside testPanelsCollapseExpand() , I first clicked the 'comments for students' panel to collapse that. And all others were as they were. 
But I am randomly getting failed test. I think issue is with `waitForPanelsToCollapse();` because on introducing some delay before `assertTrue(commentsPage.areCommentsHidden());` makes test pass consistently. Could you help me with this ?

ja```
va.lang.AssertionError: 
at org.testng.AssertJUnit.fail(AssertJUnit.java:59)
at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:24)
at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:33)
at teammates.test.cases.ui.browsertests.InstructorCommentsPageUiTest.testPanelsCollapseExpand(InstructorCommentsPageUiTest.java:167)
at teammates.test.cases.ui.browsertests.InstructorCommentsPageUiTest.allTests(InstructorCommentsPageUiTest.java:34)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:606)
at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
at org.testng.TestRunner.privateRun(TestRunner.java:767)
at org.testng.TestRunner.run(TestRunner.java:617)
at org.testng.SuiteRunner.runTest(SuiteRunner.java:334)
at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:329)
at org.testng.SuiteRunner.privateRun(SuiteRunner.java:291)
at org.testng.SuiteRunner.run(SuiteRunner.java:240)
at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
at org.testng.TestNG.run(TestNG.java:1057)
at org.testng.remote.RemoteTestNG.run(RemoteTestNG.java:111)
at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:204)
at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:175)
```

JosephineKwa commented on src/test/java/teammates/test/pageobjects/InstructorCommentsPage.java:
```diff
@@ -209,6 +209,15 @@ public void clickResponseCommentDelete(int sessionIdx, int questionIdx, int resp
     /**
+     * Clicks 'Comments for students' panel heading of the comment panel to either expand/collapse the panel body.
+     */
+    public void clickSingleCommentsPanelHeading() {
```
I think it would be better to name this method as `clickStudentCommentsPanelHeading` as `clickSingleCommentsPaneHeading` could refer to any panel

JosephineKwa commented on src/test/java/teammates/test/pageobjects/InstructorCommentsPage.java:
```diff
@@ -261,6 +270,15 @@ public void waitForPanelsToCollapse() {
     /**
+     * Waits for 'comments for students' panel to collapse.
+     */
+    public void waitForSinglePanelsToCollapse() {
```
Similarly could rename this to waitForCommentsForStudentsPanelToCollapse

JosephineKwa commented:
Hmm, I'm not getting the test failure that you mentioned. The `waitForPanelsToCollapse` should be able to correctly wait for the panels to be collapsed before allowing the test to proceed with `assertTrue(commentsPage.areCommentsHidden())`. 

However I'm getting another test failure on line 174 of `InstructorCommentsPageUiTest`. This is because the webdriver tries to click on the visibility options of a comment in the comments for students panel, which is currently hidden (due to it being collapsed in the `testPanelsCollapseExpand` method). At the start of the `testActions` method, click on the comments for students panel again to expand it so that it will be visible to the webdriver.

I think there are also still some other test templates that need to be updated, do update them as well:) 

Sorry, I wasn't able to review this earlier as I was busy this past week!

tarun14110 commented:
@JosephineKwa corrected. please review

JosephineKwa commented:
Looks good to merge, thanks for your perseverance!
